client/lib/docker.ts

```diff
@@ -0,0 +1,201 @@
+//const tar = require('tar-fs')
+//const Docker = require('dockerode');
+//const docker = new Docker({socketPath: '/var/run/docker.sock'});
+//const getPort = require('get-port');
+
+import { existsSync, createWriteStream } from 'fs'
+import * as Docker from 'dockerode';
+import * as tar from 'tar-fs';
+
+
+interface Containers{
+	core: Docker.Container,
+	worker: Docker.Container,
+	volumes: string[]
+}
+
+export class ContainerInteractor{
+	docker: Docker;
+	containerArray: Containers[];
+
+	constructor() { 
+		this.docker = new Docker({socketPath: '/var/run/docker.sock'});
+		this.containerArray = [];
+	}
+	
+	public async createCoreWorker(ports: number[]) {
+		let corePort = ports[0];
+		let workerPort = ports[1];
+		
+		let coreOpts = {
+			Image: '',
+			OpenStdin: true,
+			Env: [
+				`UDEV=0`,
+				`WORKER_TYPE=qemu`,
+			],
+			AttachStdout: true,
+			AttachStderr: true,
+			HostConfig: {
+				Privileged: true,
+				NetworkMode: `host`,
+				Mounts: Array<any>()
+			}
+		}
+
+		let workerOpts = {
+			Image: '',
+			OpenStdin: true,
+			Env: [
+				`UDEV=0`,
+				`WORKER_TYPE=qemu`,
+			],
+			AttachStdout: true,
+			AttachStderr: true,
+			Devices:[
+				{
+					PathOnHost: "/dev/net/tun",
+					PathInContainer: "/dev/net/tun"
+				},
+				{
+					PathOnHost: "/dev/kvm",
+					PathInContainer: "/dev/kvm"
+				}
+			],
+			CapAdd: [
+				"NET_ADMIN"
+			],
+			HostConfig: {
+				Privileged: true,
+				NetworkMode: `host`,
+				Mounts: Array<any>()
+			}
+		}
+		
+		// We actually need a way to generate a unique ip address to assign the virtual interface we create, to avoid conflicts when multiple jobs run on the same jenkins worker
+		let ip = (this.containerArray.length + 1).toString();
+		let env = [
+			`UDEV=0`,
+			`WORKER_TYPE=qemu`,
+			`WORKER_PORT=${workerPort}`,
+			`CORE_PORT=${corePort}`,
+			`SCREEN_CAPTURE=true`,
+			`QEMU_BRIDGE_NAME=br${corePort}`,
+			`QEMU_BRIDGE_ADDRESS=192.168.100.${ip}`,
+		]
+
+		// volume names - we make them unique to each of the containers we spawn to avoid overlap
+		let coreStorage = `core-storage-${corePort}`;
+		let reports = `reports-${corePort}`;
+		// options to mount named volumes to the containers
+		let mounts = [
+			{
+				Target:   "/data",
+				Source:   coreStorage,
+				Type:     "volume", 
+				ReadOnly: false
+			},
+			{
+				Target:   "/reports",
+				Source:   reports,
+				Type:     "volume", 
+				ReadOnly: false
+			}
+		]
+
+		// Fill in the remaining options - we do that here as we need the port name for the env variables + volume name
+		coreOpts.Env = env;
+		coreOpts.HostConfig.Mounts = mounts;
+		workerOpts.Env = env;
+		workerOpts.HostConfig.Mounts = mounts;
+
+		await this.docker.createVolume(coreStorage);
+		await this.docker.createVolume(reports);
```

Comment:
Will these fail if the volume already exists, like if the teardown was interrupted on a different run and we happen to be using the same ports on this run? I think we want to create them if they don't exist, or destroy and recreate if they do exist? Maybe reuse is fine, but not outright failure.